File: lib/text_dollar.rb
BASES = ['', 'Thousand', 'Million', 'Billion']


def text_dollar(num)
    text_num = subdivide_hundreds(num) # => Subdivide the number into hundreds for easier translations
    text_num = text_num.map {|sub_hundred| translate_hundred(sub_hundred) } # => Translates each hundred
    text_num = text_num.each_with_index.map { |word, idx| "#{word}#{BASES[idx]}"} # => Adds Base to each element
    text_num = text_num.reverse.join('') # => Creates text number
    "#{text_num}Dollars" # => Returns final dollar value
end

# Subdividing number into hundreds
# Example: 123456 => [456, 123]

def subdivide_hundreds(num)
    hundreds = Array.new
    until num == 0
        hundreds << num % 1000
        num /= 1000
    end
    hundreds
end

# Subdivides hundred value to its respective digits
# Translates each digit, and combines to create hundred number

def translate_hundred(num)
    digits = Array.new
    until num == 0
        digits << num % 10
        num /= 10
    end
    hundreds = digits[2].nil? || digits[2] == 0 ? nil : digits[2]
    tens = digits[1].nil? || digits[1] == 0 ? nil : digits[1]
    ones = digits[0].nil? || digits[0] == 0 ? nil : digits[0]

    if tens == 1
        tens = tens*10
        tens += ones if ones # => To prevent error if ones is already nil
        ones = nil # => Prevents translation of ones again
    end
    
    hundreds = "#{translate_digit(hundreds)}Hundred" if hundreds
    tens = "#{translate_tens(tens)}" if tens
    ones = "#{translate_digit(ones)}" if ones

    "#{hundreds if hundreds}#{tens if tens}#{ones if ones}"

end

# => Simple translation to word

def translate_digit(num)
    case num
    when 1
        "One"
    when 2
        "Two"
    when 3
        "Three"
    when 4
        "Four"
    when 5
        "Five"
    when 6
        "Six"
    when 7
        "Seven"
    when 8
        "Eight"
    when 9
        "Nine"
    end
end

# => Translates the tens digits
# => If the tens value ranges between 10..19, it is converted here

def translate_tens(num)
    if num >= 10
        if (16..19).include?(num)
            "#{translate_digit(num/10)}teen"
        else
            case num
            when 10
                "Ten"
            when 11
                "Eleven"
            when 12
                "Twelve"
            when 13
                "Thirteen"
            when 14
                "Fourteen"
            when 15
                "Fifteen"
            end
        end
    else
        case num
        when 2
            "Twenty"
        when 3
            "Thirty"
        when 4
            "Forty"
        when 5
            "Fifty"
        when 6
            "Sixty"
        when 7
            "Seventy"
        when 8
            "Eighty"
        when 9
            "Ninety"
        end
    end
end

File: 2017_12_12/text_dollar.js
function textDollar(num) {
    // Begin with subdividing number into 3 digits blocks
    let textNum = subDivideHundreds(num);

    // Translate each block into text
    textNum = textNum.map(function(subHundred){
        return translateHundred(subHundred);
    });

    // Add each blocks base ( thousand, million, etc )
    textNum = textNum.map(function(word, id){
        return `${word}${BASES[id]}`
    });

    // Convert array into single string in proper format
    textNum = textNum.reverse().join('');
    return `${textNum}Dollars`
}

// Subdivide number into hundreds
// Example : 123456 => [ 456, 123 ]
function subDivideHundreds(num) {
    let hundreds = [];
    while ( num > 0 ){
        hundreds.push( num % 1000 );
        num = Math.floor( num/1000 );
    }
    return hundreds;
}

// Translates number block into text
// Example: 123 => 'OneHundredTwentyThree'
function translateHundred(num) {
    let hundreds = Math.floor( num / 100 );
    let tens = Math.floor( num / 10 ) % 10;
    let ones = num % 10;
    
    if( tens === 1 ){
        tens = tens * 10 + ones;
        ones = 0;
    }

    hundreds = !!hundreds ? `${DIGITS[hundreds]}Hundred` : ``;
    tens = !!tens ? `${TENS[tens]}` : ``;
    ones = !!ones ? `${DIGITS[ones]}` : ``;

    return `${hundreds}${tens}${ones}`
}

// Set up constants

const BASES = {
    0: '',
    1: 'Thousand',
    2: 'Million',
    3: 'Billion'
};

const DIGITS = {
    1: 'One',
    2: 'Two',
    3: 'Three',
    4: 'Four',
    5: 'Five',
    6: 'Six',
    7: 'Seven',
    8: 'Eight',
    9: 'Nine'
};

const TENS = {
    2: 'Twenty',
    3: 'Thirty',
    4: 'Forty',
    5: 'Fifty',
    6: 'Sixty',
    7: 'Seventy',
    8: 'Eighty',
    9: 'Ninety',
    10: 'Ten',
    11: 'Eleven',
    12: 'Twelve',
    13: 'Thirteen',
    14: 'Fourteen',
    15: 'Fifteen',
    16: 'Sixteen',
    17: 'Seventeen',
    18: 'Eighteen',
    19: 'Nineteen'
};